code/scripting/api/objs/modelinstance.cpp

```diff
@@ -0,0 +1,323 @@
+	: _pmi(pmi)
+{}
+modelinstance_h::modelinstance_h()
+	: _pmi(nullptr)
+{}
+polymodel_instance *modelinstance_h::Get()
+{
+	return _pmi;
+}
+bool modelinstance_h::IsValid()
+{
+	return (_pmi != nullptr);
+}
+
+
+ADE_OBJ(l_SubmodelInstance, submodelinstance_h, "submodelinstance", "Submodel instance handle");
+
+submodelinstance_h::submodelinstance_h(int pmi_id, int submodel_num)
+	: _submodel_num(submodel_num)
+{
+	_pmi = model_get_instance(pmi_id);
+	_pm = _pmi ? model_get(_pmi->model_num) : nullptr;
+}
+submodelinstance_h::submodelinstance_h(polymodel_instance *pmi, int submodel_num)
+	: _pmi(pmi), _submodel_num(submodel_num)
+{
+	_pm = pmi ? model_get(pmi->model_num) : nullptr;
+}
+submodelinstance_h::submodelinstance_h()
+	: _pmi(nullptr), _pm(nullptr), _submodel_num(-1)
+{}
+polymodel_instance *submodelinstance_h::GetModelInstance()
+{
+	return IsValid() ? _pmi : nullptr;
+}
+submodel_instance *submodelinstance_h::Get()
+{
+	return IsValid() ? &_pmi->submodel[_submodel_num] : nullptr;
+}
+polymodel *submodelinstance_h::GetModel()
+{
+	return IsValid() ? _pm : nullptr;
+}
+bsp_info *submodelinstance_h::GetSubmodel()
+{
+	return IsValid() ? &_pm->submodel[_submodel_num] : nullptr;
+}
+int submodelinstance_h::GetSubmodelIndex()
+{
+	return IsValid() ? _submodel_num : -1;
+}
+bool submodelinstance_h::IsValid()
+{
+	return _pmi != nullptr && _pm != nullptr && _submodel_num >= 0 && _submodel_num < _pm->n_models;
+}
+
+
+ADE_FUNC(getModel, l_ModelInstance, nullptr, "Returns the model used by this instance", "model", "A model")
+{
+	modelinstance_h *mih;
+	if (!ade_get_args(L, "o", l_ModelInstance.GetPtr(&mih)))
+		return ade_set_error(L, "o", l_Model.Set(model_h()));
+
+	return ade_set_args(L, "o", l_Model.Set(model_h(mih->Get()->model_num)));
+}
+
+ADE_FUNC(getObject, l_ModelInstance, nullptr, "Returns the object that this instance refers to", "object", "An object")
+{
+	modelinstance_h *mih;
+	if (!ade_get_args(L, "o", l_ModelInstance.GetPtr(&mih)))
+		return ade_set_error(L, "o", l_Object.Set(object_h()));
+
+	if (mih->Get()->objnum < 0)
+		return ade_set_error(L, "o", l_Object.Set(object_h()));
+
+	return ade_set_object_with_breed(L, mih->Get()->objnum);
+}
+
+ADE_VIRTVAR(SubmodelInstances, l_ModelInstance, "modelsubmodelinstances_h", "Model submodel instances", "modelsubmodelinstances_h", "Model submodel instances, or an invalid modelsubmodelinstances handle if the model instance handle is invalid")
+{
+	modelinstance_h *mih = nullptr;
+	if (!ade_get_args(L, "o", l_ModelInstance.GetPtr(&mih)))
+		return ade_set_error(L, "o", l_ModelSubmodelInstances.Set(modelsubmodelinstances_h()));
+
+	auto pmi = mih->Get();
+	if (!pmi)
+		return ade_set_error(L, "o", l_ModelSubmodelInstances.Set(modelsubmodelinstances_h()));
+
+	if (ADE_SETTING_VAR)
+		LuaError(L, "Attempt to use Incomplete Feature: submodel instance copy");
+
+	return ade_set_args(L, "o", l_ModelSubmodelInstances.Set(modelsubmodelinstances_h(pmi)));
+}
```

Comment:
The whole modelsubmodelinstances_h thing should be a sublibrary as well, same as modelsubmodels